### DISCUSSION TITLE
Pass values between two claims of composite resources

### DISCUSSION THREAD

**Discussion by DanielRis** [crossplane/crossplane#4152] [Q&A]
Tl;dr: How do I get values from one Composite Resource Claim to another Composite Resource Claim?

I am currently trying to use crossplane to setup infrastructure in AWS. I am using the Crossplane Terraform provider to not reinvent our IAC world.

I have defined two Composite Resources, `ObjectStorage` and `OIDCRole` where the `ObjectStorage` creates an S3 bucket and some supporting resources and the `OIDCRole` creates a role that can be assumed by an K8s serviceaccount.

A user should now be able to create as many S3 buckets as they need for their specific use-case scenario, e.g.

```
apiVersion: storage.example.org/v1alpha1
kind: ObjectStorage
metadata:
  name: mybucket
  namespace: example
spec:
  compositionSelector:
    matchLabels:
      provider: aws
      service: s3
  writeConnectionSecretToRef:
    name: mybucket-bucket
---
apiVersion: storage.example.org/v1alpha1
kind: ObjectStorage
metadata:
  name: myotherbucket
  namespace: example
spec:
  compositionSelector:
    matchLabels:
      provider: aws
      service: s3
  writeConnectionSecretToRef:
    name: myotherbucket-bucket
```

Where this creates two S3 buckets, both contain the `metadata.name` in their name, including some prefix/suffix that is added by terraform. The secret this creates contains all output values in a standardized format:

```
apiVersion: v1
kind: Secret
metadata:
  name: myotherbucket-s3-storage
  namespace: example
dataString:
  name: example-mybucket-abc123
  s3_bucket_arn: arn:aws:s3:::example-mybucket-abc123
  s3_bucket_id: example-mybucket-abc123
```

Now, I want to create a IAM role that has a policy attached to allow access to these buckets. For this I want to create a IAM role that can be used by an K8s service account:
```
apiVersion: iam.example.org/v1alpha1
kind: OIDCRole
metadata:
    name: myrole
    namespace: example
spec:
    parameters:
        clusterName: "mycluster"
        description: "my description"
        oidcFullyQualifiedSubjects:
          - "system:serviceaccount:example:myserviceaccount"
        policy: |
          {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Action": [
                      "s3:GetObject",
                      "s3:PutObject",
                      "s3:DeleteObject"
                    ],
                    "Resource": [
                       "${mybucket:s3_bucket_arn}/*",
                       "${myotherbucket:s3_bucket_arn}/*"
                     ]
                },
                {
                    "Effect": "Allow",
                    "Action":
                        [
                            "kms:Decrypt",
                            "kms:GenerateDataKey"
                        ],
                    "Resource": [
                       "${mybucket:kms_arn}/*",
                       "${myotherbucket:kms_arn}/*"
                    ]
                },
            ]
          }
    compositionSelector:
        matchLabels:
            provider: aws
    writeConnectionSecretToRef:
        name: myrole-iam-role
```

Do I conceptional misunderstand the usage of Crossplane in my setup? Is it expected to create a XR out of one or multiple XRs?

**Comment by DanielRis** [comment]
Ideally (at least in my current thinking), a XR would support a function similar to the [substitution](https://fluxcd.io/flux/components/kustomize/kustomization/#post-build-variable-substitution) function of FluxCD, e.g.

```
apiVersion: iam.example.org/v1alpha1
kind: OIDCRole
metadata:
    name: myrole
    namespace: example
spec:
    parameters:
        clusterName: "mycluster"
        description: "my description"
        oidcFullyQualifiedSubjects:
          - "system:serviceaccount:example:myserviceaccount"
        policy: |
          {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Action": [
                      "s3:GetObject",
                      "s3:PutObject",
                      "s3:DeleteObject"
                    ],
                    "Resource": [
                       "${mybucket:s3_bucket_arn}/*",
                       "${myotherbucket:s3_bucket_arn}/*"
                     ]
                },
                {
                    "Effect": "Allow",
                    "Action":
                        [
                            "kms:Decrypt",
                            "kms:GenerateDataKey"
                        ],
                    "Resource": [
                       "${mybucket:kms_arn}/*",
                       "${myotherbucket:kms_arn}/*"
                    ]
                },
            ]
          }
    compositionSelector:
        matchLabels:
            provider: aws
    writeConnectionSecretToRef:
        name: myrole-iam-role
    substitueFrom:
       - kind: Secret
          name: mybucket-bucket
          as: mybucket
       - kind: Secret
          name: myotherbucket-bucket
          as: myotherbucket
```

where then the values are replaced in the resulting YAML.

**Comment by bobh66** [comment]
You can use EnvironmentConfigs to store data for one composite and retrieve it from another composite.  In your first claim/composite create an EnvironmentConfig with the data you want to share and name or label it such that the second claim/composite can find the EnvironmentConfig via name or label selection.  Then pass the required information into the second claim so it can find the EnvironmentConfig created by the first claim.
